### PR TITLE
Nguyena2/helmvalidate

### PR DESCRIPTION
**Pull request by nguyena2**
# Purpose of PR

- Research and test the automation of a helm chart validation in a Github workflow

## Type of PR

- [x] Code changes
- [x] Documentation changes
- [ ] CI-CD changes
- [ ] YAML changes

## Issues Closed or Referenced

- Closes https://github.com/retaildevcrews/wcnp/issues/1300

### CODE REVIEW

File: .github/workflows/devcontainer-build.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ on:
       - main
 
     paths:
-    - '.devcontainer/Dockerfile'
+    - 'helm-validate/Dockerfile'
```
Comment (joaquinrz) [inline .github/workflows/devcontainer-build.yaml]: I noticed that we have duplicates between .devcontainer/.devcontainer.json and helm-validate/.devcontainer/devcontainer.json do we need both?

File: .github/workflows/devcontainer-build.yaml [hunk 2]
```diff
@@ -31,7 +31,7 @@ jobs:
       - name: Build and push
         uses: docker/build-push-action@v4
         with:
-          context: '.devcontainer'
+          context: 'helm-validate/.devcontainer'
           platforms: linux/amd64,linux/arm64
           push: true
           tags: ghcr.io/joaquinrz/helmv-devcontainer:latest
```
Comment (joaquinrz) [inline .github/workflows/devcontainer-build.yaml]: we need to remove joaquinrz and friendly-fiesta and me sure that the workflows are building and pushing the images to retaildevcrews org

File: .github/workflows/integration.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            registry: ghcr.io
+            username: ${{ github.repository_owner }}
+            password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Run Kube Linter
+        id: helm-lint
+        uses: devcontainers/ci@v0.3
+        with:
+          subFolder: helm-validate
+          cacheFrom: ghcr.io/joaquinrz/helmv-devcontainer
```
Comment (joaquinrz) [inline .github/workflows/integration.yaml]: change images to retaildevcrews org

File: README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # Helm Validate - Devcontainer
 
-This repository contains a development container (devcontainer) to streamline the process of validating and testing Helm charts. The devcontainer is pre-configured with essential tools, such as `kube-linter` for linting Kubernetes resources and `helm unittest` for running Helm chart unit tests, and `kubeconform` for validating Kubernetes resources.
+This repository contains a development container (devcontainer) to streamline the process of validating and testing Helm charts. The devcontainer is pre-configured with essential tools, such as `kube-linter` for linting Kubernetes resources and `helm unittest` for running Helm chart unit tests, and `kubeconform` for validating Kubernetes resources. K3D is then used to deploy a light-weight cluster to run the Helm chart.
```
Comment (joaquinrz) [inline README.md]: not a major change but would be nice to explain that K3D is used to run the helm chart for integration testing

File: dind-k3d/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+# user args
+# some base images require specific values
+ARG USERNAME=vscode
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+
+# configure apt-get
+ENV DEBIAN_FRONTEND noninteractive
+ENV PATH $PATH:/usr/local/go/bin:/home/${USERNAME}/go/bin:/home/${USERNAME}/.local/bin:/home/${USERNAME}/.dotnet/tools:/opt/fluent-bit/bin
```
Comment (joaquinrz) [inline dind-k3d/Dockerfile]: no need for fluen-bit/bin, also we may not need .dotnet/tools

File: dind-k3d/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ENV COMPlus_EnableDiagnostics=0
+ENV COMPlus_EnableWriteXorExecute=0
+
+RUN mkdir -p /home/${USERNAME}/.local/bin && \
+    mkdir -p /home/${USERNAME}/.dotnet/tools && \
+    mkdir -p /home/${USERNAME}/.dapr/bin && \
```
Comment (joaquinrz) [inline dind-k3d/Dockerfile]: no need for .dapr and .dotnet/tools

File: dind-k3d/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ENTRYPOINT [ "/usr/local/share/docker-init.sh" ]
+CMD [ "sleep", "infinity" ]
+
+### Build k3d image from Docker-in-Docker
+FROM dind as k3d
+
+ARG USERNAME=vscode
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV PATH $PATH:/usr/local/go/bin:/usr/local/istio/bin:/home/${USERNAME}/go/bin:/home/${USERNAME}/.local/bin:/home/${USERNAME}/.dotnet/tools:/home/${USERNAME}/.dapr/bin
```
Comment (joaquinrz) [inline dind-k3d/Dockerfile]: same, no need for dapr and dotnet

File: dind-k3d/scripts/kind-k3d-debian.sh [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+fi
+
+if [ "${INSTALL_K3D}" == "true" ]; then
+    echo "Installing k3d ..."
+
+    # wget -q -O - https://raw.githubusercontent.com/rancher/k3d/main/install.sh | bash
+    wget -q -O - https://raw.githubusercontent.com/k3d-io/k3d/main/install.sh | bash
+fi
+
+# echo "Installing K9s ..."
```
Comment (joaquinrz) [inline dind-k3d/scripts/kind-k3d-debian.sh]: we can remove all these  commented code we dont use for installing k9s, flux, etc

**Review by joaquinrz — CHANGES_REQUESTED** [review]

File: .github/workflows/devcontainer-build.yaml (hunk 1 above)
Comment (nguyena2) [reply .github/workflows/devcontainer-build.yaml]: The .devcontainer.json at the root is for the person to point vs-code to open which docker/image locally. the devcontainer.jsons in both helm-validate and dind-k3d is for the github actions workflow to point towards.

File: .github/workflows/devcontainer-build.yaml (hunk 2 above)
Comment (nguyena2) [reply .github/workflows/devcontainer-build.yaml]: Updated to pushing to retaildevcrews org

File: .github/workflows/integration.yaml (hunk 3 above)
Comment (nguyena2) [reply .github/workflows/integration.yaml]: update image

File: README.md (hunk 4 above)
Comment (nguyena2) [reply README.md]: added a bit about integration testing.

File: dind-k3d/Dockerfile (hunk 5 above)
Comment (nguyena2) [reply dind-k3d/Dockerfile]: removed fluen-bit and dotnet

File: dind-k3d/Dockerfile (hunk 6 above)
Comment (nguyena2) [reply dind-k3d/Dockerfile]: removed here as well

File: dind-k3d/Dockerfile (hunk 7 above)
Comment (nguyena2) [reply dind-k3d/Dockerfile]: removed

File: dind-k3d/scripts/kind-k3d-debian.sh (hunk 8 above)
Comment (nguyena2) [reply dind-k3d/scripts/kind-k3d-debian.sh]: removed commented lines

**Review by joaquinrz — APPROVED** [review]
Approved, just need to fix new lines in a few files

**Pull request merged**